File: ru/codes/typescript/modules/ListNode.ts
/**
 * File: ListNode.ts
 * Created Time: 2022-12-10
 * Author: Justin (xiefahit@gmail.com)
 */

/* Узел связного списка */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

/* Десериализовать массив в связный список */
function arrToLinkedList(arr: number[]): ListNode | null {
    const dum: ListNode = new ListNode(0);
    let head = dum;
    for (const val of arr) {
        head.next = new ListNode(val);
        head = head.next;
    }
    return dum.next;
}

export { ListNode, arrToLinkedList };

File: ru/codes/python/modules/list_node.py
"""
File: list_node.py
Created Time: 2021-12-11
Author: krahets (krahets@163.com)
"""


class ListNode:
    """Класс узла связного списка"""

    def __init__(self, val: int):
        self.val: int = val  # Значение узла
        self.next: ListNode | None = None  # Ссылка на узел-преемник


def list_to_linked_list(arr: list[int]) -> ListNode | None:
    """Десериализовать список в связный список"""
    dum = head = ListNode(0)
    for a in arr:
        node = ListNode(a)
        head.next = node
        head = head.next
    return dum.next


def linked_list_to_list(head: ListNode | None) -> list[int]:
    """Сериализовать связный список в список"""
    arr: list[int] = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr
